### ISSUE TITLE
Make it possible to add error context

### ISSUE THREAD

**Issue by Kixunil** [toml-rs/toml#628]
To report better errors it'd be really nice to be able to add more context to the errors generated from `Deserialize` implementations. I propose adding a `context: Option<Box<dyn Any + 'static>>` field to the error and methods `set_context<T: 'static>(context: T)`, `context(&self) -> &dyn Any + 'static`, `take_context(&mut self) -> Box<dyn Any + 'static>`.

This would allow `Deserialize` implementors to downcast toml errors and provide additional custom context.

**Comment by epage** [comment]
Could you give an example of what would be inserted and how it could then be used?

**Comment by Kixunil** [comment]
Basically anything the user wishes. In my case I parse a toml file and some errors would benefit from using more precise span information (the one already in error is not good enough).

Basically, the user manually implements `Deserialize` for their type with this:

```rust
let string = String::deserialize(deserializer)?;
let bad_chars = string.chars().enumerate().filter(|(_, c)| c == '/').map(|(pos, _)| pos).collect::<Vec<_>>();
if !bad_chars.is_empty() {
    let mut error = E::custom("the string contains slashes which are forbidden");
    if let Some(toml_error) = error.downcast_mut::<toml::de::Error>() {
        toml_error.set_context(BadCharsError { bad_chars });
    }
    return Err(error);
}
```

Then when processing toml error:

```rust
match toml::from_str(s) {
    Ok(val) => process_val(val),
    Err(error) => {
        match error.context().downcast_ref::<BadCharsError>() {
            // just a demo, one might want to use codespan_reporting instad
            Some(bad_chars) => { eprintln!("Invalid chars at these positions:"); for pos in &bad_chars.bad_chars { eprintln!(" {pos}") } },
           None => eprintln!("error: {error}"),
        }
    }
}
```

(Disclaimer: I haven't tested my snippets.)

**Comment by epage** [comment]
How is that any better than including the bad chars in the custom error?

You mentioned spans but that is a whole other design space than just context and no suggestions were included regarding that.

> anything the user wishes

Except to evaluate if we're solving the right problem with the right solution, we need concrete cases.

**Comment by Kixunil** [comment]
> How is that any better than including the bad chars in the custom error?

Not being stored as string. Yeah, I could technically do that, but that's insanely ugly and not type safe. Thanks for hacky workaround while this doesn't exist though!

But maybe you mean to store `msg` of `Error::custom` as `Box<Display + Any>`? That would be great except it can't work for downcasting because of the missing `'static` bound. :(

> You mentioned spans but that is a whole other design space than just context

I mentioned it because it's my own use case. Yeah, maybe being able to override the span used in `Error` would be nicer. (But I need multiple of them!) But there's additional context information. For instance it may be easier to deserialize into `Spanned<String>` internally and post-process that than writing full blown `Visitor` etc but then it produces errors about the type not being string (yeah, my snippet shouldn't just `?` the error) which is technically true but unhelpful for the users.

Also maybe I wasn't clear enough: I do actually want to use `codespan-reporting` to get rustc-like error messages. For that I need a lot of context and for that I need to somehow pass it in error. (well, I could use TLS too but that's also horrible),

**Comment by epage** [comment]
> Also maybe I wasn't clear enough: I do actually want to use codespan-reporting to get rustc-like error messages. For that I need a lot of context and for that I need to somehow pass it in error. (well, I could use TLS too but that's also horrible),

I'm closing this issue as it is unproductive as-is.  The Issue itself is not about the actual request but is asking for piece-meal information without enough context to actually undestand the problem and discuss it.  Maintainers are not here to serve at your beck and call or to create big general solutions "just becauxse".  We need to understand what you are trying to accomplish so we can talk about solutions that will fit your actual need.,  If you want to do fancy stuff with codespan-reporting, feel free to create an issue about that use case.  This Issue is a train wreck that does not serve that purpose.

**Comment by Kixunil** [comment]
Sorry, I didn't know you perceive it as *that* bad. I thought the problem is easier to understand.

I was writing a response but during that I realized maybe I can redesign the code a lot to not need support from you at all. I'll try that first and get back if something more specific pops up.

<details>
<summary>Original response if you're curious</summary>

Just improving spans is not enough because I have multiple types with custom deserialization that can produce errors that don't fit into `serde::de::Error` trait.

If you'd like to see specific code (it's big): https://github.com/Kixunil/debcrafter/blob/master/src/input.rs
If you try to run it it has horrible error messages like "didn't match any variant of enum X" and there's always only one error. To fix that I started writing many of the `Deserialize` impls manually. (And used hacks to keep myself sane.)

So what I need is:

* Return multiple errors
* Add custom span information
* Add custom information to the messages

</details>
